src/lock.rs

```diff
@@ -0,0 +1,328 @@
+            algorithm,
+        })
+    }
+}
+
+impl std::fmt::Display for Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!(
+            "{}:{}",
+            self.algorithm,
+            hex::encode(&self.bytes)
+        ))
+    }
+}
+
+impl Serialize for Digest {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.to_string().serialize(serializer)
+    }
+}
+
+struct DigestVisitor;
+
+impl<'de> Visitor<'de> for DigestVisitor {
+    type Value = Digest;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a hexadecimal encoded cryptographic digest")
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        let mut parts = value.split(':');
+        let algorithm_tag = parts.next().ok_or(E::missing_field("algorithm"))?;
+        let algorithm = algorithm_tag
+            .parse::<DigestAlgorithm>()
+            .map_err(|_| E::custom("invalid digest algorithm"))?;
+        let bytes = parts
+            .next()
+            .ok_or(E::missing_field("bytes"))
+            .and_then(|h| hex::decode(h).map_err(|_| E::custom("invalid encoding")))?;
+        Ok(Self::Value { algorithm, bytes })
+    }
+}
+
+impl<'de> Deserialize<'de> for Digest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(DigestVisitor)
+    }
+}
+
+impl PartialEq for Digest {
+    fn eq(&self, other: &Self) -> bool {
+        self.algorithm == other.algorithm && self.bytes == other.bytes
+    }
+}
+
+impl Eq for Digest {}
+
+impl Ord for Digest {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.bytes.cmp(&other.bytes)
+    }
+}
+
+impl PartialOrd for Digest {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Captures immutable metadata about a given package
+///
+/// It is used to ensure that future installations will use the exact same dependencies.
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub digest: Digest,
+    pub registry: RegistryUri,
+    pub repository: String,
+    pub version: Version,
+    pub dependencies: Vec<PackageName>,
+    pub dependants: usize,
+}
+
+impl LockedPackage {
+    /// Captures the source, version and checksum of a Package for use in reproducible installs
+    pub fn lock(
+        package: &Package,
+        registry: RegistryUri,
+        repository: String,
+        dependants: usize,
+    ) -> Self {
+        let digest = digest::digest(&digest::SHA256, &package.tgz)
+            .try_into()
+            .unwrap(); // won't fail as SHA256 is sure to be supported
```

Comment:
I would use `.expect` here with that message.

Reply:
fair enough!